src/entities/context.cpp

```diff
@@ -0,0 +1,61 @@
+#include <veriblock/entities/context.hpp>
+
+#include "veriblock/entities/vtb.hpp"
+
+namespace altintegration {
+
+PartialVTB PartialVTB::generateFromVtb(const VTB& vtb) {
+  PartialVTB p_vtb;
+  p_vtb.endorsement = BtcEndorsement::fromContainer(vtb);
+
+  for (const auto& b : vtb.transaction.blockOfProofContext) {
+    p_vtb.btc.push_back(std::make_shared<BtcBlock>(b));
+  }
+  p_vtb.btc.push_back(std::make_shared<BtcBlock>(vtb.transaction.blockOfProof));
+
+  p_vtb.containing = std::make_shared<VbkBlock>(vtb.containingBlock);
+
+  return p_vtb;
+}
+
+std::vector<PartialVTB> PartialVTB::generateFromVtb(
+    const std::vector<VTB>& vtbs) {
+  std::vector<PartialVTB> p_vtbs;
+  p_vtbs.reserve(vtbs.size());
+
+  std::transform(vtbs.begin(),
+                 vtbs.end(),
+                 std::back_inserter(p_vtbs),
+                 [&](const VTB& vtb) -> PartialVTB {
+                   PartialVTB p_vtb = PartialVTB::generateFromVtb(vtb);
+                   for (const auto& b : vtb.context) {
+                     p_vtb.context_vbk.push_back(std::make_shared<VbkBlock>(b));
+                   }
+                   return p_vtb;
+                 });
```

Comment:
It does not look safe. When generating PartialVTBs from the vector, you populate PartialVTB.context_vbk, but when you generate PartialVTB from the single VTB, context_vbk stays uninitialised.